src/main/scala/mesosphere/marathon/core/deployment/ScalingProposition.scala

```diff
@@ -1,51 +1,56 @@
 package mesosphere.marathon
 package core.deployment
 
+import com.typesafe.scalalogging.StrictLogging
 import mesosphere.marathon.core.condition.Condition
-import mesosphere.marathon.core.instance.Instance
-import mesosphere.marathon.state.{KillSelection, Timestamp}
+import mesosphere.marathon.core.condition.Condition.UnreachableInactive
+import mesosphere.marathon.core.instance.{Goal, Instance}
+import mesosphere.marathon.state.{KillSelection, PathId, Timestamp}
 
-case class ScalingProposition(tasksToKill: Option[Seq[Instance]], tasksToStart: Option[Int])
+case class ScalingProposition(tasksToKill: Seq[Instance], tasksToStart: Option[Int])
 
-object ScalingProposition {
+object ScalingProposition extends StrictLogging {
 
   def propose(
-    runningTasks: Seq[Instance],
-    toKill: Option[Seq[Instance]],
+    instances: Seq[Instance],
+    toDecommission: Option[Seq[Instance]],
     meetConstraints: ((Seq[Instance], Int) => Seq[Instance]),
     scaleTo: Int,
-    killSelection: KillSelection): ScalingProposition = {
+    killSelection: KillSelection,
+    runSpecId: PathId): ScalingProposition = {
 
-    val killingTaskCount = runningTasks.count(_.state.condition == Condition.Killing)
+    val instancesGoalRunning: Map[Instance.Id, Instance] = instances
+      .filter(_.state.goal == Goal.Running)
+      .map(instance => instance.instanceId -> instance)(collection.breakOut)
+    val toDecommissionMap: Map[Instance.Id, Instance] = toDecommission.getOrElse(Seq.empty).map(instance => instance.instanceId -> instance)(collection.breakOut)
 
-    val runningTaskMap = Instance.instancesById(runningTasks)
-    val toKillMap = Instance.instancesById(toKill.getOrElse(Seq.empty))
-
-    val (sentencedAndRunningMap, notSentencedAndRunningMap) = runningTaskMap partition {
-      case (k, v) =>
-        toKillMap.contains(k)
+    val (sentencedAndRunningMap, notSentencedAndRunningMap) = instancesGoalRunning partition {
+      case (instanceId, instance) =>
+        toDecommissionMap.contains(instanceId) || instance.state.condition == UnreachableInactive
     }
     // overall number of tasks that need to be killed
-    val killCount = math.max(runningTasks.size - killingTaskCount - scaleTo, sentencedAndRunningMap.size)
+    val decommissionCount = math.max(instances.size - scaleTo, sentencedAndRunningMap.size)
```

Comment:
Shouldn't this be `instancesGoalRunning.size` instead of `instances.size`? Or is `instances.size == instancesGoalRunning.size - killingTaskCount`?